include/blas_quda.h

```diff
@@ -28,261 +28,308 @@ namespace quda {
       for (auto i = 0u; i < x.size(); i++) x[i].zero();
     }
 
-    inline void copy(ColorSpinorField &dst, const ColorSpinorField &src) { dst.copy(src); }
+    inline void copy(cvector_ref<ColorSpinorField> &dst, cvector_ref<const ColorSpinorField> &src)
+    {
+      for (auto i = 0u; i < src.size(); i++) { dst[i].copy(src[i]); }
+    }
 
     /**
        @brief Apply the operation y = a * x
-       @param[in] a scalar multiplier
-       @param[in] x input vector
-       @param[out] y output vector
+       @param[in] a scalar multiplier set
+       @param[in] x input vector set
+       @param[out] y output vector set
     */
-    void axy(double a, const ColorSpinorField &x, ColorSpinorField &y);
+    void axy(cvector<Complex> &a, cvector_ref<const ColorSpinorField> &x, cvector_ref<ColorSpinorField> &y);
 
     /**
        @brief Apply the rescale operation x = a * x
-       @param[in] a scalar multiplier
-       @param[in] x input vector
+       @param[in] a scalar multiplier set
+       @param[in] x set of vectors
     */
-    inline void ax(double a, ColorSpinorField &x) { axy(a, x, x); }
+    inline void ax(cvector<Complex> &a, cvector_ref<ColorSpinorField> &x) { axy(a, x, x); }
 
     /**
        @brief Apply the operation z = a * x + b * y
-       @param[in] a scalar multiplier
-       @param[in] x input vector
-       @param[in] b scalar multiplier
-       @param[in] y input vector
-       @param[out] z output vector
+       @param[in] a scalar multiplier set
+       @param[in] x input vector set
+       @param[in] b scalar multiplier set
+       @param[in] y input vector set
+       @param[out] z output vector set
     */
-    void axpbyz(double a, const ColorSpinorField &x, double b, const ColorSpinorField &y, ColorSpinorField &z);
+    void axpbyz(cvector<double> &a, cvector_ref<const ColorSpinorField> &x, cvector<double> &b,
```

Comment:
Convention choice, should we make these `const cvector<double> &a` (which I'm sure will have some recursing casting nightmares, we'll have to make things const all the way down)?

Reply:
The definition of `cvector` is
```c++
using cvector = const vector;
``` 
We only ever pass these objects around as const references to allow for temporaries to be constructed when needed, e.g., when a double scale is passed to a function that is expecting a `cvector<Complex>`.